Documentation/Evolution/DelimiterSyntax.md

```diff
@@ -0,0 +1,165 @@
+# Regular Expression Literal Delimiters
+
+- Authors: Hamish Knight, Michael Ilseman
+
+## Introduction
+
+**TODO**
+
+**TODO: Motivation for regex literals in the first place? Or is that a given?**
+
+**TODO: Overview of regex literals in other languages?**
+
+## Detailed Design
+
+A regular expression literal will be introduced using `re'...'` delimiters, within which the compiler will parse a regular expression (the details of which are outlined in **TODO: Internal syntax link**):
+
+```
+// Matches "<identifier> = <hexadecimal value>", extracting the identifier and hex number
+let regex = re'([[:alpha:]]\w*) = ([0-9A-F]+)'
+```
+
+The use of a two letter prefix allows for easy future extensibility of such literals, by allowing different prefixes to indicate different types of literal. **TODO: examples**
+
+### Parsing ambiguities
+
+The use of a single quote delimiter has a minor conflict with a couple of items of regex grammar, mainly around named groups. This includes `(?'name')`, `(?('name'))`, `\g'name'`, `\k'name'`, and `(?C'arg')`. Fortunately alternative syntax exists for all of these constructs, e.g `(?<name>)`, however we still aim to parse the single quote variants of the syntax to achieve the syntactic superset of regex grammar.
+
+To do this, a heuristic will be used when lexing a regex literal, and will check for the ending sequences `(?`, `(?(`, `\g`, `\k` and `(?C`. On encountering these, the lexer will attempt to scan ahead to the next `'` character, and then to the `'` that closes the literal. It should be noted that these are not valid regex endings, therefore this will not impact valid code.
+
+## Future Directions
+
+### Raw literals
+
+The `re'...'` syntax could be naturally extended to supporting "raw text" through allowing additional `#` characters to surround the quote characters e.g `re#'...'#`. Such literals would follow the same rules as the string literals introduced in [SE-0200].
+
+In particular:
+
+- `\` and `'` characters would become literal, e.g `re#''\n''#` expresses a regular expression pattern that literally matches against the characters `'\n'` (including the quotes).
+- Any number of `#` characters may surround the literal.
+- Escape sequences would require the same number of `#` characters as in the delimiter to be treated specially. For example, `re##'\##n'##` would be required for a newline character sequence.
+
+### Multi-line literals
+
+A natural extension to the `re'...'` syntax to support multi-line regex literals would be to allow triple quote syntax:
+
+```
+re'''
+  abc
+  def
+  '''
+```
+
+This would follow the precedent set by [SE-0168] for multi-line string literals, and obey the same rules, in particular with the stripping of any leading whitespace prior to the position of the closing delimiter.
+
```

Comment:
What about line-terminating comments that use `#`, as is typical in multi-line mode? That is, the use of multi-line delimiter is an "option" relevant to the parsing of the foreign program literal.

Would this only be a problem if the line-ending comment was immediately followed by a backslash?

Reply:
Oh, that's an interesting tidbit. I think the argument for `rx'...'` next release might be stronger in that case.

For the line-ending comments with `#`, I was wondering how that could interact with raw literals. I think it should be fairly straightforward to explain though.